mirrord/layer/src/exec.rs

```diff
@@ -125,31 +217,26 @@ pub(crate) unsafe extern "C" fn execve_detour(
     argv: *const *const c_char,
     envp: *const *const c_char,
 ) -> c_int {
-    // Do unsafe part of path conversion here.
-    let rawish_path = (!path.is_null()).then(|| CStr::from_ptr(path));
-    let mut patched_path = CString::default();
-    let final_path = patch_if_sip(rawish_path)
-        .and_then(|s| match CString::new(s) {
-            Ok(c_string) => {
-                patched_path = c_string;
-                Success(patched_path.as_ptr())
-            }
-            Err(err) => Error(Null(err)),
-        })
-        .unwrap_or(path); // Continue even if there were errors - just run without patching.
-
-    let argv_arr = Nul::new_unchecked(argv);
+    let patched_args = patch_sip_for_new_process(path, argv);
+    FN_EXECVE(patched_args.get_path(), patched_args.get_argv(), envp)
+}
 
-    // If we intercept args, we create a new array.
-    // execve takes a null terminated array of char pointers.
-    // ptr_vec will own the vector that will be passed to execve as an array.
-    let mut ptr_vec: Vec<*const c_char> = Vec::new();
-    let final_argv = intercept_tmp_dir(argv_arr)
-        .map(|new_vec| {
-            ptr_vec = new_vec; // Make sure vector still lives when we pass the pointer to execve.
-            ptr_vec.as_ptr()
-        })
-        .unwrap_or(argv);
-
-    FN_EXECVE(final_path, final_argv, envp)
+#[hook_guard_fn]
+pub(crate) unsafe extern "C" fn posix_spawn_detour(
+    pid: *const pid_t,
+    path: *const c_char,
+    file_actions: *const c_void,
+    attrp: *const c_void,
+    argv: *const *const c_char,
+    envp: *const *const c_char,
+) -> c_int {
+    let patched_args = patch_sip_for_new_process(path, argv);
```

Comment:
not sure about design/usage here - anyone who calls `patch_sip_for_new_proecss` would get the struct, and would use the methods to get fields.
I don't see a reason to have the caller have to call the external interface, since both detours have the same flow - they require path + argv, so I suggest either returning a tuple or change the fields of the struct to be `path, argv` and make those public.

Reply:
@meowjesty halp

Reply:
I'm not in tune with the SIP handling yet, so why do we need to keep these pointers alive like that anyway? I mean, these pointers are coming from C, right? So we're not really keeping them alive when we hold them (we don't get to decide when they're `free`d,).

Reply:
I'm not getting why a tuple like [this](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=d40fc021c5726d11c6727ab9e0cba981) doesn't work?

Reply:
> why do we need to keep these pointers alive like that anyway? I mean, these pointers are coming from C, right?

If we do a SIP patch, we don't use the original pointer, but create a new string (for path) or array of strings (for argv).
Example:

1. Application calls `execve` with path `/bin/bash`.
2. mirrord's `execve_detour` is executed:
  i. We check `/bin/bash` and determine it is SI-Protected.
  ii. We create  a patched binary for `/bin/bash` at `/tmp/mirrord-bin/bin/bash`.
  iii. We call the original `execve` with a pointer to the path of the patched binary. That is a String we created. If we drop that string, we're giving `execve` an invalid pointer as far as I can tell.

Reply:
> I'm not getting why a tuple like [this](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=d40fc021c5726d11c6727ab9e0cba981) doesn't work?

So we could return a 4-tuple with path pointer, optional new path string, argv pointer and optional argv vector.
This would have the same hazard as mentioned above, where it would allow destructing the tuple with `..`, thus dropping the actual objects and keeping two nice dangling pointers. Even if we wanted to ignore that problem and hope it will be ok, I do not see how returning a 4-tuple with 2 fields you must not use but also not drop is preferable over returning a struct and accessing the relevant values by a getter function.

Reply:
Alright, think I get it now.

This is kinda whats going on right now:

```rust
{
    // patched_path lifetime starts here
    let P { original_path, patched_path, .. } = patch_sip_for_new_process(path, argv);
    FN_EXECVE(patched_path.or(original_path), ...)

    // patched_path dropped
    // original_path dangling
}
```

1. You need `patched_path` to live longer?
2. Are we responsible for freeing `original_path`?

Reply:
> ```rust
>     // original_path dangling
> ```

No, that's not what's happening.

The dangling pointer scenario would only happen if we change the code to return a tuple / make the members public:

```Rust
    let (
        path_ptr, // (maybe) a raw pointer to another item of this tuple.
        argv_ptr,
        .. // path_ptr points to an object that is dropped here.
    ) = patch_sip_for_new_process(original_path_ptr, original_argv_ptr);

    FN_EXECVE(path_ptr, argv_ptr, envp_ptr); // Calling libc with the dangling pointer path_ptr.
```

Reply:
> * You need `patched_path` to live longer?

We need to pass `FN_EXECVE` raw pointers to objects we create in the detour. We get the pointers from the `patch_sip_for_new_process`. As far as I can tell, in order for those pointers to be valid, also the pointed objects need to be moved to the caller. The discussion here is about what's the best interface for that.

Reply:
> 2\. Are we responsible for freeing `original_path`?

no

Reply:
I could keep lifetime managed references to the private pointers and make them public, which would make it possible to access the pointers without calling a function, but should also prevent the struct from being destructed.

Reply:
it was decided to merge this as is and release asap and then redo it for the next release.